src/nupic/utils/MovingAverage.cpp

```diff
@@ -33,37 +33,33 @@ using namespace nupic::util;
 
 
 MovingAverage::MovingAverage(UInt wSize, const vector<Real32>& historicalValues)
-    : windowSize_(wSize)
+    : slidingWindow_(wSize, historicalValues)
 {
-  if (historicalValues.size() != 0)
-  {
-    copy(
-      historicalValues.begin() + historicalValues.size() - wSize,
-      historicalValues.end(),
-      back_inserter(slidingWindow_));
-  }
-  total_ = Real32(accumulate(slidingWindow_.begin(), slidingWindow_.end(), 0));
+ auto window = slidingWindow_.getData();
+  total_ = Real32(accumulate(begin(window), end(window), 0));
 }
 
 
-MovingAverage::MovingAverage(UInt wSize) : windowSize_(wSize), total_(0) {}
+MovingAverage::MovingAverage(UInt wSize) : 
+  slidingWindow_(wSize), total_(0) {}
 
 
 Real32 MovingAverage::compute(Real32 newVal)
 {
-  if (windowSize_ == slidingWindow_.size())
-  {
-    total_ -= slidingWindow_.front();
-    slidingWindow_.erase(slidingWindow_.begin()); // pop front element
-  }
 
-  slidingWindow_.push_back(newVal);
+  auto popped = slidingWindow_.append(newVal);
+  total_ -= popped;
   total_ += newVal;
   return getCurrentAvg();
 }
 
 
-std::vector<Real32> MovingAverage::getSlidingWindow() const
+std::vector<Real32> MovingAverage::getData() const
+{
+  return slidingWindow_.getData();
+}
+
+SlidingWindow<Real32> MovingAverage::getSlidingWindow() const
```

Comment:
The motivation came from the Likelihood PR, where I needed to access the sliding window of a moving average. 

- I agree MA should not need to expose its internal sliding window. 
- What if I need to use both sliding window & MA on the same data? It would be stupid to compute it twice: 
  - -> we need a `MovingAverage(SlidingWindow* w)` (not sure if * or &)